document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -270,18 +276,32 @@ protected String parseQueryForCompositeFilter(Filter filter) {
       case AND: {
         String childList =
             Arrays.stream(filter.getChildFilters())
-                .map(this::parseQuery)
+                .map(childFilter -> parseFilter(childFilter, paramsBuilder))
                 .filter(str -> !StringUtils.isEmpty(str))
                 .map(str -> "(" + str + ")")
                 .collect(Collectors.joining(" AND "));
         return !childList.isEmpty() ? childList : null;
       }
       default:
         throw new UnsupportedOperationException(
-            String.format("Boolean operation:%s not supported", op));
+            String.format("Query operation:%s not supported", op));
     }
   }
 
+  @VisibleForTesting
+  protected PreparedStatement buildPreparedStatement(String sqlQuery, Params params) throws SQLException {
+    PreparedStatement preparedStatement = client.prepareStatement(sqlQuery);
+    params.getStringParams().forEach((k, v) -> {
+      try {
+        // Postgres Index starts from 1
+        preparedStatement.setString(k+1, v);
```

Comment:
Since the preparedStatement allows setting different types, we should probably call the appropriate setLong/setFloat etc methods instead of assuming every param to be a String.
I think that will be functionally correct.

Reply:
The only approach I can think of, when we do [Object value = filter.getValue()](https://github.com/hypertrace/document-store/blob/f9a3b6882e239b0d05b000e9db6744b268510adb/document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java#L209), can check the datatypes `getClass`, and then consider setLong/setFloat. Please suggest if there are any alternatives?

Reply:
Yes. For now i think we should do that, until we have first class support for different types in the document-store filteer

Reply:
@avinashkolluru, tried it with `getClass()`, but the issue is querying a field in jsonb document returns a text always, so I have to cast the text to respective type(float/long/int).  Example: `SELECT * FROM mytest WHERE document->>'integer' = 123`, this doesn't work directly. Need to cast i.e. `SELECT * FROM mytest WHERE (document->>'integer')::INT = 123`. And. casting is an expensive operation in postgres, do we really want to add it?

Reply:
How do we know the type of document field type? Here, we are trying to cast based on the `input` value which can be of any type as it is provided by client. We can cast if we know the destination type or it was indicated by the interface. I think this will lead to many failure cases.

In the above e.g if the amount contains both integer and float values, would casting work if the input object is of type INT?
e.g 
```
doc1
{
'amount' : 10
}

doc2 {
'amount' : 10.5
}

Filter -> Op = Eq and Value is INT object with value 10.
```

Reply:
Also, can you verify for `>` operator for `document->>'amount'` having all int values if you use string conversation? Are you getting the expected results?
e.g 
```
doc1->amount:10, doc2->amount:5, doc3->amount:7, doc4->amount:10
select * from mytest where document->>'amount' > '4' and document->>'amount' < '8'
```
Output: `doc2, doc3`

Reply:
Verified the above with string, and works fine.

Reply:
As discussed above is a string comparison-based search. it won't work correctly on negative values as its sorting order is based on the string type.
e.g 
```
doc1->amount:-10, doc2->amount:5, doc3->amount:-5, doc4->amount:5.2, doc5->amount:-7, doc6->amount:7 doc7->amount:6
```

Also in the document store, the same field can have different types. Mongo handles it well, and it takes care during query time without throwing any exceptions. 
As an example, insert the below docs in your local Mongo
```
db.test.insertOne(
   { "name": "a", "qty": "b" }
)
db.test.insertOne(
   { "name": "b", "qty": "-5" }
)
db.test.insertOne(
   { "name": "c", "qty": "5" }
)
db.test.insertOne(
   { "name": "c", "qty": "7" }
)
```
Try queries:
```
db.test.find( {$and: [{ qty: { $gte: 'b' } }, { qty: { $lte: 'b' } }] }) - seaching all docs in collections where qty is string
db.test.find( {$and: [{ qty: { $gte: -5 } }, { qty: { $lte: 10 } }] }) - seaching all docs in collections where qty is numeric
```
So, even if we assume that there is no mixed type ingested in HT for any jsonb document for a given field, we still
need to handle numeric/boolean type queries for operator `EQ, GT, LT, GTE, LTE`.

So, as discussed, we can add the following template for `EQ, GT, LT, GTE, LTE` while parsing.
For numeric:
```
case jsonb_typeof(<field>)
WHEN ‘number’  THEN (<field>)::numeric  > ?
end
```
For boolean:
```
case jsonb_typeof(<field>)
WHEN 'boolean'  THEN (<field>)::boolean  > ?
end
```
For string (existing implementation or below for consistency) :
```
case jsonb_typeof(<field>)
WHEN 'string'  THEN (<filed>)::string  > ?
end
```